### PR TITLE
NMS-16058: Node structure union/intersection

### PR DESCRIPTION
**Pull request by synqotik**
Updated Node Structure Category filter to be either a union (resultant list contains nodes with any of the selected categories) or intersection (resultant list contains nodes with *all* of the selected categories).

Implemented as radio buttons "Any / All". Did some style tweaking to remove extra vertical margins/padding. Will have UX take a look, probably needs restyling.

Also refactored the `buildQueryParams` function and renamed component to `NodeStructurePanel` and store to `nodeStructure` instead of `hierarchy`.

Screenshot 2023-08-09 at 16 13 28

### External References

* Jira (Issue Tracker): https://opennms.atlassian.net/browse/NMS-16058

### CODE REVIEW

**Comment by RangerRick** [comment]
I don't know vue so I can't speak too much to that, but this seems reasonable. Do we have any smoke tests yet to exercise the node page?

**Review by scthompsonopen — APPROVED** [review]
I'm approving because this is just feature-branch work so my comments can be included in an upcoming sprint if desired.

My only overall comment, is the pattern for selecting items here is running a bit counter to some of the patterns I'm seeing coming over from UX normally, so you may receive this feedback from them in the future.

They've been using Chips + Autocomplete/Select relatively frequently.

image

So in the search/select box, you'd select 'Routers' 'Production' 'Test' and they'd show up as those chips with the X in them so you can remove them that way.

The way you did it definitely works, and it makes sense to me, just priming you for the feedback you MIGHT get (maybe they'll like it and we have to recode everything on our end :) )

**Comment by synqotik** [comment]
> I don't know vue so I can't speak too much to that, but this seems reasonable. Do we have any smoke tests yet to exercise the node page?

Not yet, I added a card for that.

File: ui/src/components/Nodes/utils.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import { Category, MonitoringLocation, Node, SetOperator } from '@/types'
+import { isNumber } from '@/lib/utils'
+
+const buildSearchQuery = (searchVal: string) => {
```
[the remark this answers is not in the document]
Comment (synqotik) [reply ui/src/components/Nodes/utils.ts:4]: Great idea, I added a Jira issue specifically to do unit tests.

File: ui/src/components/Nodes/NodesTable.vue [hunk 2]
```diff
@@ -235,6 +179,9 @@ const updateQuery = (val?: string) => {
   const searchQuery = buildQueryParams(val || currentSearch.value)
   const searchQueryParam: QueryParameters = { _s: searchQuery }
   const updatedParams = { ...queryParameters.value, ...searchQueryParam }
+  if (!searchQuery) {
+    delete updatedParams._s
```
[the remark this answers is not in the document]
Comment (synqotik) [reply ui/src/components/Nodes/NodesTable.vue:183]: The issue is that `queryParameters` may already contain a search query in `_s`, so we need to overwrite it if either there's a new search query, or else it was cleared. But having the `_s` with an empty string means it ends up in the query string parameters as `_s=&` or similar; it's ok but kind of messy, hence the `delete _s`, which I agree isn't great. I may refactor again to build up all the parameters and only add them as they apply.

**Pull request merged**